### PR TITLE
Implemented generate locations by an infected person

### PR DESCRIPTION
**Pull request by KohHanMing**
Implemented generate locations by infected person

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/GenerateLocationsCommand.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        List<Integer> locationIds = new ArrayList<>();
+        for (int i = 0; i < visitsByPerson.getVisitList().size(); i++) {
+            locationIds.add(visitsByPerson.getVisitList().get(i).getLocationId().getZeroBased());
+        }
+        model.updateFilteredLocationList(ModelPredicate.getPredicateShowLocationsByPerson(locationIds));
+        return new CommandResult(
+                "Generated locations for: " + model.getAddressBook()
+                        .getPersonList().get(personId.getZeroBased()).getName(),
+                false, false, CommandResult.SWITCH_TO_VIEW_ALL_LOCATIONS);
+    }
```
Comment (hopinxian) [inline src/main/java/seedu/address/logic/commands/GenerateLocationsCommand.java:68]: To abide by SLAP principle, consider abstracting out different parts into a separate methoda in the same class.
For e.g. line 50 to 62 can be abstracted into a separate method

File: src/main/java/seedu/address/logic/commands/GenerateLocationsCommand.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+ * Finds and lists all persons in address book whose name contains any of the argument keywords.
+ * Keyword matching is case insensitive.
+ */
+public class GenerateLocationsCommand extends Command {
+
+    public static final String COMMAND_WORD = "generateLocations";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Finds all locations which a person of "
+            + "the specified id (case-insensitive) visited and displays them as a list of locations.\n"
+            + "Parameters: PERSONID ...\n"
```
Comment (hopinxian) [inline src/main/java/seedu/address/logic/commands/GenerateLocationsCommand.java:27]: The "..." seems to imply that the command can accept multiple personids. Is that intended?

File: src/test/java/seedu/address/logic/commands/GenerateLocationsCommandTest.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        Index index = Index.fromOneBased(4);
+        GenerateLocationsCommand command = new GenerateLocationsCommand(index);
+        assertThrows(CommandException.class, () -> command.execute(model));
+        try {
+            command.execute(model);
+        } catch (CommandException e) {
+            assertTrue(e.getMessage().equals(expectedMessage));
+        }
+    }
+}
```
Comment (hopinxian) [inline src/test/java/seedu/address/logic/commands/GenerateLocationsCommandTest.java:93]: You have made tests that dealt with the edge cases. It would be good to have additional tests that prove the success of the command. In this case, it would be giving an infected person and confirming that the generated results are the expected results.

**Review by hopinxian — CHANGES_REQUESTED** [review]
Nearly done. Just need some more tests and abstractions.

File: src/main/java/seedu/address/model/ModelPredicate.java [hunk 4]
```diff
@@ -22,4 +23,15 @@
         person -> person.getInfectionStatus().getStatusAsBoolean();
     public static final Predicate<Person> PREDICATE_SHOW_ALL_QUARANTINED =
         person -> person.getQuarantineStatus().getStatusAsBoolean();
+
+    /** {@code Predicate} to handle location list */
```
Comment (siangernlow) [inline src/main/java/seedu/address/model/ModelPredicate.java:27]: The JavaDocs for this Predicate could be clearer as this Predicate is not as straightforward as the others.

File: src/main/java/seedu/address/model/ModelPredicate.java [hunk 5]
```diff
@@ -22,4 +23,15 @@
...
+    public static Predicate<Location> getPredicateShowLocationsByPerson(List<Integer> locationIds) {
+        return location -> {
+            boolean condition = location.getId().getZeroBased() == locationIds.get(0);
+            for (int i = 1; i < locationIds.size(); i++) {
+                condition = condition || location.getId().getZeroBased() == locationIds.get(i);
```
Comment (siangernlow) [inline src/main/java/seedu/address/model/ModelPredicate.java:32]: Might be better to name the conditions separately with informative names as this line of code can get confusing.

**Review by siangernlow — APPROVED** [review]
LGTM, but please take note of the comments in the ModelPredicate class.

**Comment by hopinxian** [comment]
Pull request was merged using Admin powers without going through 2 approvals. 
However, noted that I would have given approval after the recent commits, so no harm done. Please be careful in the future though.
